core/kernel/msg_param.c

```diff
@@ -0,0 +1,221 @@
+		case OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_TMEM_INOUT:
+			if (pages_cnt >= num_params)
+				return -1;
+			if (!(params[i].attr & OPTEE_MSG_ATTR_FRAGMENT) &&
+			    (i != num_params - 1))
+				return -1;
+			pages[pages_cnt++] = params[i].u.tmem.buf_ptr
+				& ~SMALL_PAGE_MASK;
+			break;
+		default:
+			return -1;
+		}
+	}
+	return pages_cnt;
+}
+
+/**
+ * msg_param_map_buffer() - map parameters buffer into OP-TEE VA space
+ * @pa_params - physical pointer to parameters
+ * @num_params - number of parameters
+ *
+ * return:
+ *  struct shmem_mapping of mapped buffer on success
+ *  NULL on error.
+ */
+struct mobj *msg_param_map_buffer(paddr_t pa_params, size_t num_params)
+{
+	struct mobj *mobj = NULL;
+	struct optee_msg_param *params;
+	paddr_t *pages = NULL;
+	size_t pages_cnt = 1;
+	size_t max_pages;
+	size_t args_size;
+	size_t i;
+
+	args_size = OPTEE_MSG_GET_ARG_SIZE(num_params);
+	max_pages = args_size / SMALL_PAGE_SIZE + 1;
+
+	pages = calloc(max_pages, sizeof(paddr_t));
+	if (!pages)
+		goto err;
+
+	pages[0] = pa_params & ~SMALL_PAGE_MASK;
+	mobj = mobj_mapped_shm_alloc(pages, 1, 0, 0);
+	if (!mobj)
+		goto err;
+
+	params = mobj_get_va(mobj, pa_params & SMALL_PAGE_MASK);
+
+	for (i = 0; i < num_params; i++) {
+		if ((params->attr & OPTEE_MSG_ATTR_TYPE_MASK) ==
+		    OPTEE_MSG_ATTR_TYPE_NEXT_FRAGMENT) {
+			if (pages_cnt >= max_pages)
+				goto err;
+
+			pages[pages_cnt] = params->u.tmem.buf_ptr;
+			mobj_free(mobj);
+
+			mobj = mobj_mapped_shm_alloc(pages + pages_cnt,
+						     1, 0, 0);
+			if (!mobj)
+				goto err;
+
+			pages_cnt++;
+			params = mobj_get_va(mobj, 0);
+		} else {
+			params++;
+			/* Check if we are not overflowing over page */
+			if (((vaddr_t)params & SMALL_PAGE_MASK) == 0)
+				goto err;
+		}
+	}
+
+	mobj_free(mobj);
+
+	mobj = mobj_mapped_shm_alloc(pages, pages_cnt, 0, 0);
+	if (!mobj)
+		goto err;
+
+	free(pages);
+	return mobj;
+err:
+	free(pages);
+	mobj_free(mobj);
+	return NULL;
+}
+
+
+/**
+ * msg_param_init_memparam() - fill memory reference parameter for RPC call
+ * @param	- parameter to fill
+ * @mobj	- mobj describing the shared memory buffer
+ * @offset	- offset of the buffer
+ * @size	- size of the buffer
+ * @cookie	- NW cookie of the shared buffer
+ * @dir		- data direction
+ *
+ * Idea behind this function is that thread_rpc_alloc() can return
+ * either buffer from preallocated memory pool, of buffer constructed
+ * from supplicant's memory. In first case parameter will have type
+ * OPTEE_MSG_ATTR_TYPE_TMEM_* and OPTEE_MSG_ATTR_TYPE_RMEM_ in second case.
+ * This function will fill parameter structure with right type, depending on
+ * the passed mobj.
+ *
+ * return:
+ *	true on success, false on failure
+ */
+bool msg_param_init_memparam(struct optee_msg_param *param, struct mobj *mobj,
+			     size_t offset, size_t size,
+			     uint64_t cookie, enum msg_param_mem_dir dir)
+{
+	if (mobj_matches(mobj, CORE_MEM_NON_SEC)) {
```

Comment:
I was testing this pull request without any changes in the other gits. I was expecting everything to just work falling back to using the predefined shm pool, especially since this log entry:
```
ERROR:   TEE-CORE:tee_entry_exchange_capabilities:113: No NS DDR configured for the platform. Disabling dynamic SHM
```
But it seems this test still causes the param to be initialized with registered SHM causing a failure later.